### ISSUE TITLE
Calling plot() when using a tensor as inference source returns a floating point array

### ISSUE THREAD

**Issue by banoffee-pie** [ultralytics/ultralytics#4295]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Detection

### Bug

Hesitant to call it a bug, but this does seem like it is unintended behaviour(?)

If you provide a tensor of type float32, values in range [0,1] and shape BCHW as an inference source, calling `plot()` returns a floating point ndarray which if you try to follow [the instructions in the docs](https://docs.ultralytics.com/modes/predict/#probs) retulsts in the following below error when converting it to a PIL image

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /opt/conda/lib/python3.10/site-packages/PIL/Image.py:3070, in fromarray(obj, mode)
   3069 try:
-> 3070     mode, rawmode = _fromarray_typemap[typekey]
   3071 except KeyError as e:

KeyError: ((1, 1, 3), '<f4')

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
Cell In[73], line 1
----> 1 display_result(wrapper_results[0])

Cell In[56], line 3, in display_result(results_obj)
      1 def display_result(results_obj: Results):
      2     im_array = results_obj.plot()  # plot a BGR numpy array of predictions
----> 3     im = Image.fromarray(im_array[..., ::-1])  # RGB PIL image
      4     im.show()

File /opt/conda/lib/python3.10/site-packages/PIL/Image.py:3073, in fromarray(obj, mode)
   3071     except KeyError as e:
   3072         msg = "Cannot handle this data type: %s, %s" % typekey
-> 3073         raise TypeError(msg) from e
   3074 else:
   3075     rawmode = mode

TypeError: Cannot handle this data type: (1, 1, 3), <f4
```

For it to work, the output of `plot()` has to be cast to uint8.

### Suggested fix:
In the plot function, if the original image is a tensor, it is multiplied by 255 but never cast to uint8. I believe that wrapping the whole thing in an `.astype(numpy.uint8)` is the intended behaviour, as this is the type that a PIL image converted to a numpy array would have had.

https://github.com/ultralytics/ultralytics/blob/c9be1f3cce89778f79fb462797b8ca0300e3813d/ultralytics/engine/results.py#L217-L218

### Minimal Reproducible Example

```python
import torch
import numpy
from ultralytics import YOLO

pytorch_yolo = YOLO("yolov8n")

# converts image located at path to a PyTorch Tensor
def open_and_preprocess(path: str):
    pil_img = Image.open(path).convert("RGB")

    # resize to what the model expects
    size = pytorch_yolo.model.args["imgsz"]
    resized_img = pil_img.resize((size, size))

    np_arr = numpy.array(resized_img)
    np_chw = np_arr.transpose([2, 0, 1]) # transpose from hwc to chw
    np_batch = numpy.expand_dims(np_chw, 0) # add batch dimension
    np_float = np_batch.astype(numpy.single) / 255 # convert to float
    return torch.from_numpy(numpy.ascontiguousarray(np_float))

# display results
def display_result(results_obj: Results):
    im_array = results_obj.plot()  # plot a BGR numpy array of predictions
    im = Image.fromarray(im_array[..., ::-1])  # RGB PIL image
    im.show()  # show image

image = open_and_preprocess("/path/to/image");
results = pytorch_yolo(image)
display_results(results)

```

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@banoffee-pie thanks for the bug report! I'm able to reproduce with a slightly updated code example:

```python
import numpy
from ultralytics import YOLO
from PIL import Image
pytorch_yolo = YOLO("yolov8n")
# converts image located at path to a PyTorch Tensor
def open_and_preprocess(path: str):
    pil_img = Image.open(path).convert("RGB")
    # resize to what the model expects
    size = pytorch_yolo.model.args["imgsz"]
    resized_img = pil_img.resize((size, size))
    np_arr = numpy.array(resized_img)
    np_chw = np_arr.transpose([2, 0, 1]) # transpose from hwc to chw
    np_batch = numpy.expand_dims(np_chw, 0) # add batch dimension
    np_float = np_batch.astype(numpy.single) / 255 # convert to float
    return torch.from_numpy(numpy.ascontiguousarray(np_float))
# display results
def display_result(results_obj):
    im_array = results_obj[0].plot()  # plot a BGR numpy array of predictions
    im = Image.fromarray(im_array[..., ::-1])  # RGB PIL image
    im.show()  # show image
image = open_and_preprocess("ultralytics/assets/bus.jpg")
results = pytorch_yolo(image)
display_result(results)

```

**Comment by glenn-jocher** [comment]
@banoffee-pie great news 😃! Your original issue may now be resolved ✅ in PR #4297. 

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!
